### PR TITLE
[serialize] Add a fast unchecked tag decoder 

### PR DESCRIPTION
**Pull request by ryanhall07**
This adds an unchecked tag decoder/iterator which removes all safety
checks (i.e ref checking) to improve the performance. It is safe to use
this when the underlying bytes are not shared and are safe to reuse.

A good example of this is the downsampler, that is single threaded and
has to match a metric id against many rules. This significantly
increases the performance of the downsampler.

### CODE REVIEW

File: src/x/serialize/unchecked_decoder.go [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+	}
+	if len(d.currentTagName) == 0 {
+		return ErrEmptyTagNameLiteral
+	}
+
+	d.currentTagValue, err = d.decodeID()
+	if err != nil {
+		return err
+	}
+	if err != nil {
```
Comment (robskillington) [inline src/x/serialize/unchecked_decoder.go:121]: This seems repeated?

File: src/x/serialize/unchecked_decoder.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+		return nil, err
+	}
+
+	if limit := d.maxTagLength; l > limit {
+		return nil, fmt.Errorf("tag literal too long [ limit = %d, observed = %d ]", limit, int(l))
+	}
+
+	if len(d.data) < int(l) {
+		return nil, errInvalidByteStreamIDDecoding
```
Comment (robskillington) [inline src/x/serialize/unchecked_decoder.go:139]: Is it valid to have a tag with zero value length? If so this logic breaks. 

I think technically we allow that but we probably shouldn’t, although something perhaps to check against the other decoder.

**Review by robskillington — APPROVED** [review]
LGTM other than minor comments, yay

File: src/x/serialize/unchecked_decoder.go (hunk 2 above)
Comment (robskillington) [reply src/x/serialize/unchecked_decoder.go:139]: Ah, looks like this is the same 👍 
https://github.com/m3db/m3/blob/2b93a5e159a9f2005b2f904d777ff7123637edda/src/x/serialize/decoder.go#L155-L157

Should retain same logic then 👍

File: src/x/serialize/unchecked_decoder.go (hunk 1 above)
Comment (ryanhall07) [reply src/x/serialize/unchecked_decoder.go:121]: lol whoops. good catch

File: src/x/serialize/unchecked_decoder.go (hunk 2 above)
Comment (ryanhall07) [reply src/x/serialize/unchecked_decoder.go:139]: lol yea I copied that original decoder to start :)

File: src/x/serialize/unchecked_decoder.go (hunk 1 above)
Comment (ryanhall07) [reply src/x/serialize/unchecked_decoder.go:121]: surprised the linter doesn't catch this

**Pull request merged**